### DISCUSSION TITLE
Finding files in sub-folders by name

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#2032] [Q&A] [answered]
I have a certain need, and I'm wondering whether there are existing VisiData commands that can help with it. 

I'm in a `DirSheet` which has a lot of folders in it. Each of these folders has in it a file called `rollout.jsonl`. I want to go through these files one-by-one. I wish there was a way to say "In this `DirSheet`, please search for files with this name in all subfolders recursively, and open something like a `DirSheet` where each row is one of these files, so I could open each of them by pressing Enter."

Is there something like that?

**Comment by saulpw** [comment, marked as answer]
Use the `-r`/`--dir-recurse` flag.  Select all `rollout.jsonl` files.  `"` to pull them into their own sheet.  Or `g Enter` to open them up, each in their own sheet.

**Reply by cool-RR** [reply]
Interesting, but the decision to view those rollout files is something I make when I'm already inside VisiData, so a flag isn't a good solution for me. I'm thinking that the best move for me might be to write a `DirSheet` command that takes a regex pattern and opens a new `DirSheet` with just those files. Do you think that'll work?

**Reply by saulpw** [reply]
Well any flag is also an option, so you could similarly set the option, or create a command that sets the option and reloads the dirsheet.  I think this would be the simplest way, but up to you of course.

**Reply by cool-RR** [reply]
Oh, I keep making the mistake of only searching the command list and not searching the options list. Thank you.

**Reply by cool-RR** [reply]
I tried to write that command now:

```
@BaseSheet.api
def rfind(dir_sheet: DirSheet, regex: str) -> None:
    pattern = re.compile(regex)
    new_dir_sheet = DirSheet(str(dir_sheet.source), dir_sheet.source)
    vd.push(new_dir_sheet)
```

It pushes a new `DirSheet`, but it's empty of files, and its source appears to be `None`. What am I doing wrong?

**Reply by saulpw** [reply]
use `source=dir_sheet.source`.  All non-keyword args are components of the name.

**Reply by cool-RR** [reply]
Thank you. I'm now trying to do this: `vd.unhide_cols(new_dir_sheet, ('directory',), dir_sheet.visibleRows)`

It's failing with `TypeError: unhide_cols() takes 3 positional arguments but 4 were given`.

What am I doing wrong?

**Reply by saulpw** [reply]
Well the definition of unhide_cols is `def unhide_cols(vd, cols, rows):` and you're passing in the sheet, which it doesn't take as a parameter.

**Reply by cool-RR** [reply]
Right, silly mistake, thank you.

**Reply by cool-RR** [reply]
Now I've got this code:

```
@BaseSheet.api
def rfind(dir_sheet: DirSheet, regex: str) -> None:
    new_dir_sheet = DirSheet(source=dir_sheet.source)
    new_dir_sheet.options.dir_recurse = True
    (directory_column,) = (column for column in new_dir_sheet.columns
                           if column.name == 'directory')
    vd.unhide_cols((directory_column,), dir_sheet.visibleRows)
    
    new_dir_sheet.selectByIdx(vd.searchRegex(new_dir_sheet, regex=regex,
                                             columns=(directory_column,)))

    vd.push(new_dir_sheet)
```

That `new_dir_sheet.selectByIdx` line causes this for some reason: 

```
 Traceback (most recent call last):
   File "/home/ramrachum/.local/lib/python3.10/site-
 packages/visidata/threads.py", line 198, in _toplevelTryFunc
     t.status = func(*args, **kwargs)
   File "/home/ramrachum/.local/lib/python3.10/site-
 packages/visidata/selection.py", line 74, in select
     for r in (Progress(rows, 'selecting') if progress else rows):
   File "/home/ramrachum/.local/lib/python3.10/site-
 packages/visidata/selection.py", line 97, in <genexpr>
     self.select((self.rows[i] for i in rowIdxs), progress=False)
   File "/home/ramrachum/.local/lib/python3.10/site-
 packages/visidata/search.py", line 47, in searchRegex
     c = findMatchingColumn(sheet, sheet.rows[rowidx], columns, regex.search)
 IndexError: list index out of range
```

Any idea why?

**Reply by saulpw** [reply]
The sheet `reload()` doesn't start working until after the push, so there are no rows at the point you're trying the `selectByIdx`.  Moreover, it's done in its own thread, so even if you change the ordering in `rfind`, it won't do what you want.  The general tactic is to do the work in the sheet's reload, or to wait until the sheet is done loading (with `vd.sync()` if necessary).  But this is now getting into much more complicated territory, so again, it might be simpler (a lot simpler) to use the existing commands/options.

As a side-note, you can do `sheet.column('directory').width = 10` to unhide the 'directory' column.

**Reply by cool-RR** [reply]
That's disappointing. Thanks for the answer.

**Reply by cool-RR** [reply]
I've been looking at this issue again, because I do want this feature. I found `FileListSheet` which sounds close to what I want. I tried to find where it's used, but apparently it's used only in `open_fdir`, which doesn't seem to be called anywhere. Is this dead code? Should I delete it?

**Reply by cool-RR** [reply]
And do you think that subclassing `DirSheet` and overriding `iterload` to only load files that match a certain regex would be a good way to implement the desired feature here?

**Reply by saulpw** [reply]
> I've been looking at this issue again, because I do want this feature. I found FileListSheet which sounds close to what I want. I tried to find where it's used, but apparently it's used only in open_fdir, which doesn't seem to be called anywhere. Is this dead code? Should I delete it?

The `fdir` filetype is so you can do something like:

`find . -name rollout.jsonl | vd -f fdir`

This sounds like it might what you want.  I forgot about this feature, my apologies!

**Reply by saulpw** [reply]
Also FYI `vd.open_foo` functions aren't dead code, they are what's called to create a sheet when the filetype is given with `-f foo`.

**Reply by cool-RR** [reply]
Great.

I wrote this: 

```
@BaseSheet.api
def rfind(dir_sheet: DirSheet, regex: str) -> None:
    from visidata.shell import FileListSheet
    paths = [path for path in pathlib.Path(dir_sheet.source).rglob('*')
             if re.search(regex, path.name)]
    file_list_sheet = FileListSheet(name='rfind', source=paths)
    vd.push(file_list_sheet)
```

But I get this: 

```
     for fn in self.source.open():
 AttributeError: 'list' object has no attribute 'open'
```

I'm guessing I'm passing that list of paths wrong. How should I pass it?

**Reply by cool-RR** [reply]
Oh, I got it: 

```python
@BaseSheet.api
def rfind(dir_sheet: DirSheet, regex: str) -> None:
    from visidata.shell import FileListSheet
    import visidata
    
    paths = (path for path in pathlib.Path(dir_sheet.source).rglob('*')
             if re.search(regex, path.name))
    fptext = iter(f'{path}\n' for path in paths)
    file_list_sheet = FileListSheet(name='rfind', source=visidata.Path('foo', fptext=fptext))
    vd.push(file_list_sheet)
    vd.unhide_cols((file_list_sheet.column("directory"),), file_list_sheet.visibleRows)
        
DirSheet.addCommand(
    None,
    'rfind',
    'rfind(regex=input("find recursively by filename regex: ", type="regex", '
                      'defaultLast=True))',
    'Find files by name regex recursively'
)
```

It works. You know, the way that VisiData is built internally is so weird, and the `RepeatFile` object inside `Path` is a prime example of that. Question: What is the `given` argument to `Path`? What is the significance of that string?

**Reply by saulpw** [reply]
`given` is the url/path that was given to the constructor, verbatim.  So `.`, not expanded, or the url, unparsed.

As for RepeatFile inside Path, what's the non-weird way of caching a file-like so that an HTTP stream can be opened multiple times without making multiple requests?

**Reply by cool-RR** [reply]
> As for RepeatFile inside Path, what's the non-weird way of caching a file-like so that an HTTP stream can be opened multiple times without making multiple requests?

I assume you're inviting constructive criticism here.

Whatever's going on with the processing of file names out of an HTTP stream shouldn't have anything to do with the sheet that displays a list of files. Your process for managing that HTTP stream should output an iterator of paths (which could come as either strings or `Path` objects), which then `FileListSheet` should consume. I figured out by trial-and-error that you designed it so the stream of paths should be strings, with newlines at the end for some reason, and then they need to be fed into the `fptext` of a `Path` object for some unclear reason, and that `Path` needs some string for its `given` attribute which the docs don't say what its significance is (and your reply didn't help because you described it from an implementation perspective rather than a user's / product perspective, i.e. I have no idea what actually is done with this `given`.)

Here's a saner API:

```
paths = (path for path in pathlib.Path(dir_sheet.source).rglob('*')
         if re.search(regex, path.name))
file_list_sheet = FileListSheet(name='rfind', paths=paths)
```

If you don't want to change the signature of `__init__` you could also do:

```
file_list_sheet = FileListSheet.from_paths(name='rfind', paths=paths)
```


By the way, even though in my code above I initialize the `FileListSheet` with `name='rfind'`, its name shows up as `''`. Any idea why?

**Reply by saulpw** [reply]
> By the way, even though in my code above I initialize the FileListSheet with name='rfind', its name shows up as ''. Any idea why?

Name is not a kwarg; for a few versions, it's been a list of name components as `*args`.  This API may be confusing, maybe it should be `name_or_names` and the constructor can figure out how to deal with it whether it's a single string name or a list of name components.

**Reply by cool-RR** [reply]
Works now, thank you. I also changed it to a more reasonable `f'{dir_sheet}_rfind'`.

**Reply by saulpw** [reply]
Ah, this is the purpose of those `*args` names!  Call it as `FileListSheet(*dir_sheet.names, 'rfind', ...)` and then there will be two name components, which will look just like your `{dir_sheet}_rfind`, but will be broken down into name components.  This isn't immediately useful, but may well factor into how the sheet tabs look on the status bar very shortly.

**Comment by cool-RR** [comment]
That makes sense, I changed my code accordingly now. Thank you.
